cinder/volume/api.py

```diff
@@ -2245,11 +2245,16 @@ def attachment_create(self,
                                                   volume_ref,
                                                   instance_uuid)
         if connector:
-            connection_info = (
-                self.volume_rpcapi.attachment_update(ctxt,
-                                                     volume_ref,
-                                                     connector,
-                                                     attachment_ref.id))
+            try:
+                connection_info = (
+                    self.volume_rpcapi.attachment_update(ctxt,
+                                                         volume_ref,
+                                                         connector,
+                                                         attachment_ref.id))
+            except Exception:
+                with excutils.save_and_reraise_exception():
+                    self.attachment_delete(ctxt, attachment_ref)
```

Comment:
This theoretically could result in an attachment being put into error_detaching state if the driver's detach_volume() fails due to the volume being in a weird state due to failed attach.   But the vmdk driver doesn't implement detach_volume() so this is probably ok.